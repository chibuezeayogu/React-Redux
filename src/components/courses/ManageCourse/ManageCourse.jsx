import React, { Component, Fragment } from 'react';
import { array, func, shape } from 'prop-types';
import { connect } from 'react-redux';
import { withRouter, Prompt } from 'react-router-dom';

import inputValidation from '../../../helpers/inputValidation';
import { saveCourse } from '../../../actions/courseActions';
import CourseForm from '../CourseForm/CourseForm';
import getById from '../../../helpers/getCourseById';
import formattedAuthorsDropdown from '../../../helpers/formatAuthorsDropdown';

/**
 * @class ManageCoursePage
 * @extends Component
 */
export class ManageCourse extends Component {
	static initialState = () => ({
		course: { 
			id: '',
			title: '',
			length: '',
			authorId: '',
			category: '',
			watchHref: '',
		},
		hasUnsavedChanges: false,
		loading: false,
		errors: {}
	})
	
	static getDerivedStateFromProps(nextProps, prevState) {
		if (nextProps.course.id && !prevState.course.id) {
			return {
				...ManageCourse.initialState(),
				course: nextProps.course
			};
		}
		
		return null;
	}
	
	state = ManageCourse.initialState();

	/**
   * 
   * @memberof ManageCoursePage
   * @param {Object} event
   * @returns {void}
   */
	onChange = (event) => {
		const { errors, course } = this.state;
	
		if (Object.entries(errors).length > 0) {
			this.setState({ errors: {} });
		}
		course[event.target.name] = event.target.value;
		
		this.setState({ 
			course,
			hasUnsavedChanges: true
		});
	}
  
	/**
   * 
   * @memberof ManageCoursePage
   * @param {Object} evt
   * @returns {void}
   */
	saveCourse = (evt) => {
		evt.preventDefault();
		const err = inputValidation(this.state.course);
		if (!err.isEmpty) {
			this.setState({ errors: err.errors });
		} else {
			this.setState({ loading: true });
			this.props.saveCourse(this.state.course)
				.then(() => {
					this.setState({ 
						hasUnsavedChanges: false,
						loading: false
					});
					this.props.history.push('/courses');
				});
		}
	}

	render() {
		const { authors } = this.props;
		const {
			hasUnsavedChanges, course, loading, errors 
		} = this.state;

		return (
			<Fragment>
				<Prompt
					when={hasUnsavedChanges}
					message="Your unsaved data will be lost"
				/>
				<div className="jumbotron">
					<h1 className="display-4">Manage Course</h1>
				</div>
				<CourseForm
					allAuthors={authors}
					onChange={this.onChange}
					onSave={this.saveCourse}
					course={course}
					errors={errors}
					loading={loading}/>
			</Fragment>
		);
	}
}

ManageCourse.propTypes = {
	course: shape({}).isRequired,
	authors: array.isRequired,
	saveCourse: func.isRequired
};


const mapStateToProps = ({ coursesReducer, authorsReducer }, props) => {
	const { id } = props.match.params;
	let course = ManageCourse.initialState();

	if (id && coursesReducer.courses.length > 0) {
		course = Object.assign({}, getById(coursesReducer.courses, id));
	}

	return {
		course,
		authors: formattedAuthorsDropdown(authorsReducer.authors)
	};
};

export default
connect(mapStateToProps,
	{ saveCourse }
)(ManageCourse);
