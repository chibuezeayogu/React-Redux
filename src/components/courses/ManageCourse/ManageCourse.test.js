import React from 'react';
import { mount } from 'enzyme';

import { ManageCourse, ManageCoursePage } from './ManageCourse';

describe('ManageCourse Component', () => {
	const props = {
		history: {
			push: jest.fn()
		},
		match: { params: { id: 'courseID' } },
		courses: [{
			id: 'courseID', 
			watchHref: '', 
			title: '', 
			authorId: '', 
			length: '', 
			category: ''
		},
		{
			id: '', 
			watchHref: '', 
			title: '', 
			authorId: '', 
			length: '', 
			category: ''
		}],
		saveCourse: jest.fn(),
		authors: [{
			id: 'cory-house',
			firstName: 'Cory',
			lastName: 'House'
		}]
	};
 
	const state = {
		course: {
			watchHref: 'http://www.pluralsight.com', 
			title: 'title', 
			authorId: 'cory-house', 
			length: '5:40', 
			category: 'Software Architecture'
		},
		auhtors: [],
		errors: {}
	};
  
	const wrapper = mount(<ManageCourse {...props} {...state}/>);

	it('should render correctly', () => {
		expect(wrapper).toMatchSnapshot();
	});

	it('should call onChange method', () => {
		const Spy = jest
			.spyOn(wrapper.instance(), 'onChange');
		const event = { 
			preventDefault: jest.fn(),
			target: {
				value: '',
				name: ''
			}
		};
		wrapper.setState({ errors: { id: '*required' } });
		wrapper.instance().onChange(event);
		expect(Spy).toHaveBeenCalled();
	});

	it('should call saveCourse method empty fields', () => {
		const Spy = jest
			.spyOn(wrapper.instance(), 'saveCourse');
		const event = { preventDefault: jest.fn() };
		wrapper.instance().saveCourse(event);
		expect(Spy).toHaveBeenCalled();
	});

	it('should call saveCourse method with valid input', () => {
		const Spy = jest
			.spyOn(wrapper.instance(), 'saveCourse');
		const event = { preventDefault: jest.fn() };
		wrapper.setProps({ match: { params: { id: '' } } });
		wrapper.setState({
			course: {
				id: 'course-id',
				watchHref: 'http://www.pluralsight.com', 
				title: 'title', 
				authorId: 'cory-house', 
				length: '5:40', 
				category: 'Software Architecture'
			}
		});
		wrapper.instance().saveCourse(event);
		expect(Spy).toHaveBeenCalled();
	});
});
