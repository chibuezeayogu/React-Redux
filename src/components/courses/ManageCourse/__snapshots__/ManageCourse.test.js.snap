// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageCourse Component should render correctly 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(ManageCourse)
        auhtors={Array []}
        authors={
          Array [
            Object {
              "firstName": "Cory",
              "id": "cory-house",
              "lastName": "House",
            },
          ]
        }
        course={
          Object {
            "authorId": "cory-house",
            "category": "Software Architecture",
            "length": "5:40",
            "title": "title",
            "watchHref": "http://www.pluralsight.com",
          }
        }
        courses={
          Array [
            Object {
              "authorId": "",
              "category": "",
              "id": "courseID",
              "length": "",
              "title": "",
              "watchHref": "",
            },
            Object {
              "authorId": "",
              "category": "",
              "id": "",
              "length": "",
              "title": "",
              "watchHref": "",
            },
          ]
        }
        errors={Object {}}
        history={
          Object {
            "push": [MockFunction],
          }
        }
        match={
          Object {
            "params": Object {
              "id": "courseID",
            },
          }
        }
        saveCourse={[MockFunction]}
      >
        <ManageCourse
          auhtors={Array []}
          authors={
            Array [
              Object {
                "firstName": "Cory",
                "id": "cory-house",
                "lastName": "House",
              },
            ]
          }
          course={
            Object {
              "authorId": "cory-house",
              "category": "Software Architecture",
              "length": "5:40",
              "title": "title",
              "watchHref": "http://www.pluralsight.com",
            }
          }
          courses={
            Array [
              Object {
                "authorId": "",
                "category": "",
                "id": "courseID",
                "length": "",
                "title": "",
                "watchHref": "",
              },
              Object {
                "authorId": "",
                "category": "",
                "id": "",
                "length": "",
                "title": "",
                "watchHref": "",
              },
            ]
          }
          errors={Object {}}
          history={
            Object {
              "push": [MockFunction],
            }
          }
          match={
            Object {
              "params": Object {
                "id": "courseID",
              },
            }
          }
          saveCourse={[Function]}
        >
          <Prompt
            message="Your unsaved data will be lost"
            when={false}
          />
          <div
            className="jumbotron"
          >
            <h1
              className="display-4"
            >
              Manage Course
            </h1>
            <hr
              className="my-4"
            />
          </div>
          <CourseForm
            allAuthors={
              Array [
                Object {
                  "text": "Cory House",
                  "value": "cory-house",
                },
              ]
            }
            course={
              Object {
                "authorId": "",
                "category": "",
                "id": "courseID",
                "length": "",
                "title": "",
                "watchHref": "",
              }
            }
            errors={Object {}}
            loading={false}
            onChange={[Function]}
            onSave={[Function]}
          >
            <form>
              <TextInput
                label="Title"
                name="title"
                onChange={[Function]}
                value=""
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="title"
                  >
                    Title
                     
                  </label>
                  <div
                    className="field"
                  >
                    <input
                      className="form-control"
                      name="title"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </TextInput>
              <SelectInput
                defaultOption="Select Author"
                label="Author"
                name="authorId"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "text": "Cory House",
                      "value": "cory-house",
                    },
                  ]
                }
                value=""
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="authorId"
                  >
                    Author
                  </label>
                  <div
                    className="field"
                  >
                    <select
                      className="form-control"
                      name="authorId"
                      onChange={[Function]}
                      value=""
                    >
                      <option
                        value=""
                      >
                        Select Author
                      </option>
                      <option
                        key="cory-house"
                        value="cory-house"
                      >
                        Cory House
                      </option>
                    </select>
                  </div>
                </div>
              </SelectInput>
              <TextInput
                label="Category"
                name="category"
                onChange={[Function]}
                value=""
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="category"
                  >
                    Category
                     
                  </label>
                  <div
                    className="field"
                  >
                    <input
                      className="form-control"
                      name="category"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </TextInput>
              <TextInput
                label="Length"
                name="length"
                onChange={[Function]}
                value=""
              >
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="length"
                  >
                    Length
                     
                  </label>
                  <div
                    className="field"
                  >
                    <input
                      className="form-control"
                      name="length"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </TextInput>
              <button
                className="btn btn-primary"
                disabled={false}
                onClick={[Function]}
              >
                Save
              </button>
            </form>
          </CourseForm>
        </ManageCourse>
      </Connect(ManageCourse)>
    </Provider>
  </Router>
</BrowserRouter>
`;
